lib/Sema/CSDiag.cpp

```diff
@@ -5410,13 +5410,75 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
 
     return true;
   }
-  
-  // If we found an exact match, this must be a problem with a conversion from
-  // the result of the call to the expected type.  Diagnose this as a
-  // conversion failure.
-  if (calleeInfo.closeness == CC_ExactMatch)
+
+  // We have found exact much for the arguments, which means this is either
+  // generic or conversion failure, let's try to diagnose that.
+  if (calleeInfo.closeness == CC_ExactMatch) {
+    // This is not an initializer, we don't know what might be wrong.
+    if (!isa<TypeExpr>(fnExpr))
+      return false;
+
+    // We have an initializer expression for some type (might be generic),
```

Comment:
I'm not sure knowing something is a TypeExpr is enough to assume it's an initializer. What if it's followed by `.self`?